### 100 Numpy Exercises

[numpy-100](https://github.com/rougier/numpy-100)中文版及一些注解。

#### 1. 导入Numpy

In [1]:
import numpy as np

#### 2. 打印Numpy版本号及其配置

In [2]:
np.__version__

'2.2.2'

In [3]:
np.show_config()

Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /opt/_internal/cpython-3.13.0/lib/python3.13/site-packages/scipy_openblas64/include
    lib directory: /opt/_internal/cpython-3.13.0/lib/python3.13/site-packages/scipy_openblas64/lib
    name: scipy-openblas
    openblas configuration: OpenBLAS 0.3.28  USE64BITINT DYNAMIC_ARCH NO_AFFINITY
      Haswell MAX_THREADS=64
    pc file directory: /project/.openblas
    version: 0.3.28
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /opt/_internal/cpython-3.13.0/lib/python3.13/site-packages/scipy_openblas64/include
    lib directory: /opt/_internal/cpython-3.13.0/lib/python3.13/site-packages/scipy_openblas64/lib
    name: scipy-openblas
    openblas configuration: OpenBLAS 0.3.28  USE64BITINT DYNAMIC_ARCH NO_AFFINITY
      Haswell MAX_THREADS=64
    pc file directory: /project/.openblas
    version: 0.3.28
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    na

#### 3. 创建一个长度(size)为10的向量(vector)

注意这里就是创建一个`np.ndarray`

In [4]:
z = np.zeros(10)
z

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

#### 4. 计算数组的内存大小

In [5]:
z = np.zeros((10, 10))
print("%d bytes" % (z.size * z.itemsize))

800 bytes


注意这里就是简单的把总元素个数`z.size`（100）， 乘上单个元素所占的内存`z.itemsize`(8 bytes).

#### 5. 在命令行打印出Numpy中add函数的帮助文档信息

这里主要是有关命令行调用Python的问题，我们可以在命令行从`python --help`开始， 找到`python -c`符合我们的要求，所以这里只需要在命令行执行`python -c "import numpy; numpy.info(numpy.add)"`即可。就等同于在Python解释器中执行如下程序：
```python
import numpy
numpy.info(numpy.add)
```

此外，针对Numpy,我们可以有很多种方式查看文档:上面的`np.info(np.add)`, 以及利用help的`help(np.add)`, 和比较少用的doc方法的调用`print(np.add.__doc__)`（输出和`np.info`一致）, 一般来说用`np.info`就可以了，也比较方便。如果使用`IPython`的话，可以直接`np.add?`回车

#### 6. 创建一个长度为10的向量，其第五个值为1， 其他为0

In [6]:
z = np.zeros(10)
z[4] = 1
z

array([0., 0., 0., 0., 1., 0., 0., 0., 0., 0.])

#### 7. 创建一个包含从10到49所有整数的向量

In [7]:
z = np.arange(10, 50)
z

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26,
       27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43,
       44, 45, 46, 47, 48, 49])

注意Python也有内建函数`range`具有相似的功能，相对而言，Numpy的`arange`由于使用了内存优化技术，其效率要高很多。我们可以做个简单的水平对比。例子来自*Scipy Lecture Notes*

In [8]:
%timeit [i**2 for i in range(1000)]

35.8 μs ± 686 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [9]:
%timeit a = np.arange(1000) ** 2

1.24 μs ± 11.7 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


#### 8. 反转一个向量（逆序）

In [10]:
z = np.arange(10)
z

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [11]:
z = z[::-1]
z

array([9, 8, 7, 6, 5, 4, 3, 2, 1, 0])

#### 9. 创建一个3x3的矩阵， 包含数字0到8

In [12]:
z = np.arange(0, 9).reshape(3, 3)
z

array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])

这里可以在原题目的基础上进行拓展，上面的实现中0到8是可以看作以行为顺序，如我我们希望0到8以列为顺序排列呢？

In [13]:
# 列为顺序
z = np.arange(0, 9).reshape(3, 3).T
z

array([[0, 3, 6],
       [1, 4, 7],
       [2, 5, 8]])

其实只需要把原来的矩阵转置就可以了:-)

#### 10. 找出[1,2,0,0,4,0] 中非0数字的位置

In [14]:
z = np.array([1,2,0,0,4,0])
z.nonzero()

(array([0, 1, 4]),)

#### 11. 创建3x3的单位矩阵

In [15]:
z = np.eye(3)
z

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

#### 12. 创建3x3x3数组，以随机数字填充

In [16]:
z = np.random.random((3, 3, 3))
z

array([[[0.91269293, 0.19869989, 0.10494772],
        [0.94254533, 0.54957537, 0.18154804],
        [0.62087253, 0.03436994, 0.10061259]],

       [[0.46483655, 0.04538607, 0.24690816],
        [0.61331421, 0.87302408, 0.86804378],
        [0.29061092, 0.08830448, 0.49009229]],

       [[0.73823004, 0.31329852, 0.00544942],
        [0.66911608, 0.28970614, 0.10625876],
        [0.0564424 , 0.69529167, 0.60116431]]])

#### 13. 创建10x10数组，以随机数字填充，并找出其中的最大值和最小值

In [17]:
z = np.random.random((10, 10))
z

array([[0.14177117, 0.00729834, 0.3577473 , 0.5093345 , 0.33809338,
        0.968723  , 0.69471792, 0.64509132, 0.53671914, 0.956913  ],
       [0.14036251, 0.87562729, 0.20146337, 0.15422012, 0.38561734,
        0.9832719 , 0.42683887, 0.16495553, 0.20418151, 0.36398929],
       [0.38192358, 0.76909836, 0.61049641, 0.70675359, 0.37924232,
        0.2771104 , 0.83215307, 0.49363321, 0.7639181 , 0.0843528 ],
       [0.67606865, 0.48380313, 0.41508288, 0.77187434, 0.54709549,
        0.76973074, 0.4801813 , 0.58027311, 0.0714549 , 0.3682779 ],
       [0.51959408, 0.76206898, 0.82688925, 0.15712273, 0.89316505,
        0.73019538, 0.88622424, 0.56651193, 0.68346814, 0.32424828],
       [0.3768179 , 0.93308506, 0.44728135, 0.51397597, 0.36283366,
        0.19783704, 0.9120004 , 0.40290882, 0.82498654, 0.94539743],
       [0.67460675, 0.96833451, 0.66571123, 0.81026997, 0.95125769,
        0.56538073, 0.44341773, 0.16423002, 0.20933498, 0.28797144],
       [0.66605841, 0.83987406, 0.1208704

In [18]:
z.max(), z.min()

(np.float64(0.9916244922895215), np.float64(0.007298338277469174))

#### 14. 创建长度为10的随机向量，并计算其均值

In [19]:
z = np.random.random((10))
z

array([0.22209021, 0.22987129, 0.75781618, 0.22587786, 0.2619966 ,
       0.56671689, 0.42444305, 0.97267822, 0.44521316, 0.83585921])

In [20]:
z.mean()

np.float64(0.49425626609907863)

#### 15. 创建一个二维数组，其边界值为1，内部值为0

In [21]:
z = np.ones((5, 5))
z[1:-1, 1:-1] = 0
z

array([[1., 1., 1., 1., 1.],
       [1., 0., 0., 0., 1.],
       [1., 0., 0., 0., 1.],
       [1., 0., 0., 0., 1.],
       [1., 1., 1., 1., 1.]])

#### 16. 将现有的数组（nxn）用0组成的边界包裹

In [22]:
z = np.ones((5, 5))
z

array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

In [23]:
m = np.pad(z, (1, 1), mode='constant', constant_values=0)
m

array([[0., 0., 0., 0., 0., 0., 0.],
       [0., 1., 1., 1., 1., 1., 0.],
       [0., 1., 1., 1., 1., 1., 0.],
       [0., 1., 1., 1., 1., 1., 0.],
       [0., 1., 1., 1., 1., 1., 0.],
       [0., 1., 1., 1., 1., 1., 0.],
       [0., 0., 0., 0., 0., 0., 0.]])

#### 17. 下列表达式的结果是什么
```python
0 * np.nan
np.nan == np.nan
np.inf > np.nan
np.nan - np.nan
np.nan in set([np.nan])
0.3 == 3 * 0.1
```

In [24]:
0 * np.nan

nan

有`np.nan`参与的算术操作返回均为`np.nan`

In [25]:
np.nan == np.nan

False

这里是合理的，比如我们从数据集读出两列数据全部是`np.nan`， 如果上面的表达式设计为返回`True`， 
那么我们在完全不知道两列数据的情况下就判定二者是相等的，这显然是不合理的。所以这里返回的是`False`.

In [26]:
np.inf > np.nan

False

同样地，我们不能比较无穷大与缺失值的大小

In [27]:
np.nan in set([np.nan])

True

In [28]:
0.3 == 3 * 0.1

False

由于浮点数（`float`）运算存在误差，我们不能直接比较其大小。Numpy为我们提供了`np.allclose`函数来比较浮点数之间的近似相等。此外，此函数还支持`np.ndarray`的比较。

In [29]:
np.allclose(0.3, 3 * 0.1)

True

#### 18. 创建一个5x5的矩阵，其中1，2，3，4正好在矩阵对角线元素下方

In [30]:
z = np.diag(np.arange(1, 5), k=-1)
z

array([[0, 0, 0, 0, 0],
       [1, 0, 0, 0, 0],
       [0, 2, 0, 0, 0],
       [0, 0, 3, 0, 0],
       [0, 0, 0, 4, 0]])

#### 19. 创建一个8x8的矩阵，并用0，1标记为国际象棋棋盘的形式
如下所示, 黑色部分标记为1.
<img src="https://timgsa.baidu.com/timg?image&quality=80&size=b9999_10000&sec=1564734957&di=10e4e0640ee6888baf9593b9cc877934&imgtype=jpg&er=1&src=http%3A%2F%2Fpic.51yuansu.com%2Fpic3%2Fcover%2F02%2F10%2F13%2F599f0c01a035b_610.jpg" alt="drawing" width="200"/>

In [31]:
z = np.zeros((8, 8))
z[1::2, ::2] = 1  # 第2, 4, 6, 8行填充
z[::2, 1::2] = 1  # 第1， 3， 5， 7行填充
z

array([[0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.],
       [0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.],
       [0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.],
       [0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.]])

#### 20. 现有维度为6x7x8的数组，找出其中第100个元素的索引(x, y, z)

In [32]:
print(np.unravel_index(99, (6, 7, 8)))

(np.int64(1), np.int64(5), np.int64(3))


上面的是给出的答案，一开始我并不知道这个函数，采用了下面的方法，可以作为参考。

In [33]:
z = np.arange(6*7*8).reshape(6, 7, 8)
np.where(z == 99)

(array([1]), array([5]), array([3]))

这是通过Numpy找出来具体位置，但是具体计算的方法并未给出，这里简单解释下。

首先，我们可以形象第考虑“数组的维度越往后，对应数据的颗粒度越小”，也就是说，在上面的例子中，我们可以认为6x7x8的立方体是通过如下的方法来构建的：先将所有的一列值(6 * 7 * 8)排成一行，之后每8个组成一个“长条”， 这样就有6*7个长条；之后将每7个长条，上下拼接，铺成一个平面；这样我们就有6个平面，将这6个平面堆起来，就得到了我们最终的“立方体”。

那么第100个元素又在哪里呢？为方便起见，我们从“颗粒度”大的开始，依次定位其位置。首先，可以知道每一层含有7*8=56个元素，所以由`100 // 56 = 1`得其位于第二层， 对应到该维度得到索引就是1，即返回的`array[1]`。之后在第二层中继续定位， 去除第一层的56个元素，这里还剩下44个。又由于平面为7x8的，所以由`44 // 8 = 5`得其位于第6行，对应该维度的索引是5， 即返回的`array[5]`, 最后剩下4个元素在新的一行，对应维度的索引为3， 即返回的`array[3]`.由此得到最终的索引为（1， 5， 3）

#### 21. 用`tile`函数创建一个8x8的棋盘

In [34]:
unit = np.array([[0, 1], [1, 0]])
z = np.tile(unit, (4, 4))
z

array([[0, 1, 0, 1, 0, 1, 0, 1],
       [1, 0, 1, 0, 1, 0, 1, 0],
       [0, 1, 0, 1, 0, 1, 0, 1],
       [1, 0, 1, 0, 1, 0, 1, 0],
       [0, 1, 0, 1, 0, 1, 0, 1],
       [1, 0, 1, 0, 1, 0, 1, 0],
       [0, 1, 0, 1, 0, 1, 0, 1],
       [1, 0, 1, 0, 1, 0, 1, 0]])

`tile`的原意就是铺瓷砖，是其作用的一个形象的比喻，这里我们把8x8的棋盘划分为4x4=16块“瓷砖”（这里的`unit`）， 之后将其平铺在一起即可。

#### 22. 标准化一个5x5的随机矩阵

In [35]:
z = np.random.random((5, 5))
z = (z - z.mean()) / z.std()
z

array([[-0.98946773,  0.66702557,  0.88334071, -0.04070042,  0.24049612],
       [-0.91960843,  0.56040709,  1.38309019,  0.06136736,  1.13726585],
       [ 1.23967812, -1.91054732,  0.72464328,  0.80027137,  0.51918724],
       [-0.8007358 , -0.15011298,  0.46203467, -0.77730673,  0.3091174 ],
       [-2.087569  ,  1.1161328 ,  0.47500233, -1.70083596, -1.20217574]])

#### 23. 创建一个自定义的包含四个无符号字节（RGBA）的`dtype`来描述颜色

In [36]:
color = np.dtype([("r", np.ubyte, 1),
                  ("g", np.ubyte, 1),
                  ("b", np.ubyte, 1),
                  ("a", np.ubyte, 1)])
color

dtype([('r', 'u1', (1,)), ('g', 'u1', (1,)), ('b', 'u1', (1,)), ('a', 'u1', (1,))])

#### 24. 5x3的矩阵与3x2的矩阵相乘

In [37]:
z = np.dot(np.ones((5, 3)), np.ones((3, 2)))
z

array([[3., 3.],
       [3., 3.],
       [3., 3.],
       [3., 3.],
       [3., 3.]])

In [38]:
# 也可以使用操作符 @
z = np.ones((5, 3)) @ np.ones((3, 2))
z

array([[3., 3.],
       [3., 3.],
       [3., 3.],
       [3., 3.],
       [3., 3.]])

#### 25. 给定一个一维数组，将值在3和8之间的数字变为其负数

In [39]:
z = np.arange(10)
z[(z > 3) & (z < 8)] *= -1
z

array([ 0,  1,  2,  3, -4, -5, -6, -7,  8,  9])

#### 26. 下面脚本的输出是什么

```python
# Author: Jake VanderPlas

print(sum(range(5),-1))
from numpy import *
print(sum(range(5),-1))
```

In [40]:
sum(range(5), -1)

9

这里是使用Python内置的`sum`函数, 它把所有的参数都当作求和的一部分相加， 这里就是简单地将所有的数字相加，10 - 1 = 9

In [41]:
np.sum(range(5), -1)

np.int64(10)

这里使用的是Numpy中的`np.sum`， 这里的`-1`并非待加的数字，而是另外一个参数的值，代表多维数组在求和时各个轴求和的顺序。具体可以`help(np.sum)`

#### 27. `z`是整数组成的向量，判断下列表达式是否正确

In [42]:
z = np.arange(10)
z

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [43]:
# 1
z ** z

array([        1,         1,         4,        27,       256,      3125,
           46656,    823543,  16777216, 387420489])

前面提到过， 对于数组之间的`**`等算术运算， 是元素一一对应进行运算的(element-wise), 如这里387420489，就等于9**9

In [44]:
# 2
2 << z >> 2

array([  0,   1,   2,   4,   8,  16,  32,  64, 128, 256])

本质上进行两次移位运算，也就是等于`(2 << z) >> 2`.下面将其拆开来看。

In [45]:
2 << z

array([   2,    4,    8,   16,   32,   64,  128,  256,  512, 1024])

就是将2分别左移0， 1， 2， ...， 9位，得到的就是`2 << 0, 2<<1, ..., 2<<9`,如下所示：

In [46]:
part1 = [2 << i for i in range(10)]
part1

[2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]

可以看到与`2 << z`的输出是一致的。

之后就是进行右移位操作，不同之处在于，这里是对于数组`2 << z`中的每个元素进行右移位，分别右移2个位置。每个数值右移2代表对每个值`x`，取`x // 4`, 对`part1`继续处理

In [47]:
[i // 4 for i in part1]

[0, 1, 2, 4, 8, 16, 32, 64, 128, 256]

可以看到其与`2 << z >> 2`的输出是一致的。

In [48]:
# 3
z <- z

array([False, False, False, False, False, False, False, False, False,
       False])

这里涉及的主要是优先级的问题，随便找个操作符，如`<`,通过`help("<")`即可查看所有操作符的优先级，默认是从低优先级到高优先级。可以看到，`-1`相比`<`具有更高的优先级， 所以这里就等同于`z < (-z)`，测试如下

In [49]:
z < (-z)

array([False, False, False, False, False, False, False, False, False,
       False])

In [50]:
# 4
1j * z

array([0.+0.j, 0.+1.j, 0.+2.j, 0.+3.j, 0.+4.j, 0.+5.j, 0.+6.j, 0.+7.j,
       0.+8.j, 0.+9.j])

对复数的运算的支持

In [51]:
# 5
z/1/1

array([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.])

也就是`(z/1)/1`

In [52]:
# 6
try:
    z < z > z
except Exception as e:
    print(e)

The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()


这里参考了下Python表达式的[文档](https://docs.python.org/2/reference/expressions.html#not-in),和[stackoverflow](https://stackoverflow.com/questions/18755059/is-abc-valid-python)
找到
>Formally, if `a, b, c, …, y, z` are expressions and `op1, op2, …, opN` are comparison operators, then a `op1 b op2 c ... y opN z` is equivalent to `a op1 b and b op2 c and ... y opN z`, except that each expression is evaluated at most once.

就是说，在一个表达式里面进行连续比较的时候，如`x < y <= z`,首先是符合语法的，其等同于`x < y adn y <=z`, 只不过对于重复的元素（这里的`y`）只估计一次。所以我们在Python原生的`list`中进行上述`z`的运算是可以正常返回的，代码如下。

In [53]:
l1 = [1, 2]
l1 < l1 > l1

False

In [54]:
(l1 < l1) and (l1 > l1)

False

因为`l1 < l1`和`l1 > l1`全部是`False`,所以其`and`也是`False`.但是对于我们的`z`，也就是`np.ndarray`类型的数据，情况就有所不同。

这里`z < z > z`依旧是估计为`z < z and z > z`, 其中`z < z`与`z > z`都是可以正常返回的，且结果都是一个长度为`z.size()`的`array`， 元素全部是布尔值。

In [55]:
z < z

array([False, False, False, False, False, False, False, False, False,
       False])

In [56]:
z > z

array([False, False, False, False, False, False, False, False, False,
       False])

不可行的是两者之间的`and`。 因为在进行`and`操作时，Numpy无法确切地知道形如`array([False, False, ...])`的数组到底是估计为`False`，还是`True`， 因为这里有两种方法来定义一个数组的布尔值：其一是`ａｌｌ`，即所有的元素全是`Ｔｒｕｅ`才判定为`Ｔｒｕｅ`，否则为`Ｆａｌｓｅ`；　另外一种方法是`ａｎｙ`，　即只要数组中有一个`Ｔｒｕｅ`，我们就判定其为`Ｔｒｕｅ`, 否则判定为`Ｆａｌｓｅ`. 正是这种不确定性使得Numpy报错，并建议使用`any`或者`all`.

#### 28. 下列表达式的结果是什么

```python
np.array(0) / np.array(0)
np.array(0) // np.array(0)
np.array([np.nan]).astype(int).astype(float)
```

In [57]:
np.array(0) / np.array(0)

/tmp/ipykernel_70371/873513115.py:1: RuntimeWarning: invalid value encountered in divide
  np.array(0) / np.array(0)


np.float64(nan)

返回`nan`并带有警告说在进行真除(true_divide)的时候出现问题，即0做分母。

In [58]:
np.array(0) // np.array(0)

/tmp/ipykernel_70371/2018018105.py:1: RuntimeWarning: divide by zero encountered in floor_divide
  np.array(0) // np.array(0)


np.int64(0)

返回0, 并带有警告说在进行地板除(floor_divide)的时候出现问题，即0做分母。

In [59]:
np.array([np.nan]).astype(int).astype(float)

/tmp/ipykernel_70371/699728972.py:1: RuntimeWarning: invalid value encountered in cast
  np.array([np.nan]).astype(int).astype(float)


array([-9.22337204e+18])

#### 29. 舍入浮点数数组，使其尽可能远离0
即-0.3， -0.5， -0.6等近似为-1，而非0； 0.3， 0.5， 0.6等近似为1，而非0.

In [60]:
z = np.random.uniform(-10, 10, 10)
z

array([-9.10814698, -6.66448921, -9.38987006,  1.70882624, -2.88238723,
        4.33997806,  1.78832019, -1.84208733,  9.45841641,  7.59649495])

In [61]:
np.copysign(np.ceil(np.abs(z)), z)

array([-10.,  -7., -10.,   2.,  -3.,   5.,   2.,  -2.,  10.,   8.])

#### 30. 找到两个数组中相同的元素

~我们首先考虑内置的函数，但是我们不知道是否有类似的函数，所以我们可以灵活使用`np.lookfor`来找出我们要的函数.~

`np.lookfor` was removed in the NumPy 2.0 release. Search NumPy's documentation directly.

In [62]:
np.info(np.intersect1d)

Find the intersection of two arrays.

Return the sorted, unique values that are in both of the input arrays.

Parameters
----------
ar1, ar2 : array_like
    Input arrays. Will be flattened if not already 1D.
assume_unique : bool
    If True, the input arrays are both assumed to be unique, which
    can speed up the calculation.  If True but ``ar1`` or ``ar2`` are not
    unique, incorrect results and out-of-bounds indices could result.
    Default is False.
return_indices : bool
    If True, the indices which correspond to the intersection of the two
    arrays are returned. The first instance of a value is used if there are
    multiple. Default is False.

Returns
-------
intersect1d : ndarray
    Sorted 1D array of common and unique elements.
comm1 : ndarray
    The indices of the first occurrences of the common values in `ar1`.
    Only provided if `return_indices` is True.
comm2 : ndarray
    The indices of the first occurrences of the common values in `ar2`.
    Only provided if 

根据文档就可以直接使用了。

In [63]:
z1 = np.arange(-5, 5)
z2 = np.arange(10)
np.intersect1d(z1, z2)

array([0, 1, 2, 3, 4])

#### 31. 如何忽视所有Numpy的警告（不推荐）

In [64]:
# 自杀模式启动:-)
defaults = np.seterr(all="ignore")
Z = np.ones(1) / 0
# 恢复理智
_ = np.seterr(**defaults)

In [65]:
# 也可以定义错误处理的细节
with np.errstate(divide='warn'):
    Z = np.ones(1) / 0

/tmp/ipykernel_70371/1125215274.py:3: RuntimeWarning: divide by zero encountered in divide
  Z = np.ones(1) / 0


#### 32. 下面的表达式会返回`True`吗
```python
np.sqrt(-1) == np.emath.sqrt(-1)
```

In [66]:
np.sqrt(-1) == np.emath.sqrt(-1)

/tmp/ipykernel_70371/244602691.py:1: RuntimeWarning: invalid value encountered in sqrt
  np.sqrt(-1) == np.emath.sqrt(-1)


np.False_

In [67]:
np.sqrt(-1), np.emath.sqrt(-1)

/tmp/ipykernel_70371/178168484.py:1: RuntimeWarning: invalid value encountered in sqrt
  np.sqrt(-1), np.emath.sqrt(-1)


(np.float64(nan), np.complex128(1j))

#### 33. 如何获取昨天，今天，明天的日期

In [68]:
yesterday = np.datetime64('today', 'D') - np.timedelta64(1, 'D')
today     = np.datetime64('today', 'D')
tomorrow  = np.datetime64('today', 'D') + np.timedelta64(1, 'D')

In [69]:
yesterday, today, tomorrow

(np.datetime64('2025-01-22'),
 np.datetime64('2025-01-23'),
 np.datetime64('2025-01-24'))

#### 34. 如何获取2016年7月全部31天的日期

In [70]:
z = np.arange('2016-07', '2016-08', dtype='datetime64[D]')
z

array(['2016-07-01', '2016-07-02', '2016-07-03', '2016-07-04',
       '2016-07-05', '2016-07-06', '2016-07-07', '2016-07-08',
       '2016-07-09', '2016-07-10', '2016-07-11', '2016-07-12',
       '2016-07-13', '2016-07-14', '2016-07-15', '2016-07-16',
       '2016-07-17', '2016-07-18', '2016-07-19', '2016-07-20',
       '2016-07-21', '2016-07-22', '2016-07-23', '2016-07-24',
       '2016-07-25', '2016-07-26', '2016-07-27', '2016-07-28',
       '2016-07-29', '2016-07-30', '2016-07-31'], dtype='datetime64[D]')

#### 35. 如何以替换的方式（in place）计算$((A+B)*(-A/2))$（不通过复制）

In [71]:
A = np.ones(3)*1
B = np.ones(3)*2
C = np.ones(3)*3
np.add(A,B,out=B)
np.divide(A,2,out=A)
np.negative(A,out=A)
np.multiply(A,B,out=A)

array([-1.5, -1.5, -1.5])

#### 36. 用5种方法提取随机数组中的整数部分

In [72]:
z = np.random.uniform(0, 10, 10)
z

array([5.51372669, 5.04763111, 3.35317182, 9.39849161, 2.40559731,
       8.26336304, 3.50529748, 5.14460516, 3.06226711, 5.56110297])

In [73]:
# 1
z - z % 1

array([5., 5., 3., 9., 2., 8., 3., 5., 3., 5.])

In [74]:
# 2
np.floor(z)

array([5., 5., 3., 9., 2., 8., 3., 5., 3., 5.])

In [75]:
# 3
np.ceil(z) - 1

array([5., 5., 3., 9., 2., 8., 3., 5., 3., 5.])

In [76]:
# 4
z.astype(int)

array([5, 5, 3, 9, 2, 8, 3, 5, 3, 5])

In [77]:
# 5
np.trunc(z)

array([5., 5., 3., 9., 2., 8., 3., 5., 3., 5.])

#### 37. 创建一个5x5的矩阵，每行均为0到4

In [78]:
# 1,答案的方法
z = np.zeros((5, 5))
z += np.arange(5)
z

array([[0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.],
       [0., 1., 2., 3., 4.]])

In [79]:
# 2,利用tile
z = np.tile(np.arange(5), (5, 1))
z

array([[0, 1, 2, 3, 4],
       [0, 1, 2, 3, 4],
       [0, 1, 2, 3, 4],
       [0, 1, 2, 3, 4],
       [0, 1, 2, 3, 4]])

#### 38. 现有一个可以生成10个整数的生成器函数，利用其建立一个数组

In [80]:
# 1, 答案的方法
def gen():
    for i in range(10):
        yield i
        
z = np.fromiter(gen(), dtype=float, count=-1)
z

array([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.])

In [81]:
# 2, 列表解析
z = np.array([i for i in gen()])
z

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

#### 39. 创建一个长度为10，范围从0到1的向量（不包括0，1）

In [82]:
z = np.linspace(0, 1, 11, endpoint=False)[1:]
z

array([0.09090909, 0.18181818, 0.27272727, 0.36363636, 0.45454545,
       0.54545455, 0.63636364, 0.72727273, 0.81818182, 0.90909091])

#### 40. 创建一个长度为10的随机数组并排序

In [83]:
z = np.random.random(10)
z

array([0.28352028, 0.39968781, 0.54395735, 0.73555733, 0.58550852,
       0.27926301, 0.31790326, 0.00846767, 0.13790058, 0.28961763])

In [84]:
z.sort()
z

array([0.00846767, 0.13790058, 0.27926301, 0.28352028, 0.28961763,
       0.31790326, 0.39968781, 0.54395735, 0.58550852, 0.73555733])

#### 41. 对于长度较小的数组，如何更高效地求和（相对`np.sum`）

In [85]:
z = np.arange(10)
np.add.reduce(z)

np.int64(45)

In [86]:
%timeit np.add.reduce(z)

843 ns ± 5.59 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [87]:
%timeit np.sum(z)

1.61 μs ± 12.5 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


可以看到`np.add.reduce`此时差不多快上一倍

#### 42. 检查两个数组A， B是否相等

In [88]:
A = np.random.randint(0,2,5)
B = np.random.randint(0,2,5)

In [89]:
# 1， 已知A，B的shape相等
# 存在容错，适用于浮点数的比较
np.allclose(A, B)

False

In [90]:
# 2. 同时检查shape与数值
# 要求数值完全相等
np.array_equal(A, B)

False

#### 43. 限制数组为不可变数组（read only）

In [91]:
z = np.zeros(10)
z.flags

  C_CONTIGUOUS : True
  F_CONTIGUOUS : True
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False

In [92]:
z.flags.writeable = False
z.flags

  C_CONTIGUOUS : True
  F_CONTIGUOUS : True
  OWNDATA : True
  WRITEABLE : False
  ALIGNED : True
  WRITEBACKIFCOPY : False

In [93]:
try:
    z[0] = 1
except Exception as e:
    print(e)

assignment destination is read-only


#### 44. 给定10x2矩阵代表平面座标系中座标，将其转化为极座标系座标

In [94]:
z = np.random.random((10, 2))
x, y = z[:, 0], z[:, 1]
r = np.sqrt(x**2 + y**2)
theta = np.arctan2(y, x)
r, theta

(array([0.87241029, 0.10307188, 0.95993714, 0.89020759, 1.20696461,
        0.55103234, 0.7218472 , 1.19380778, 0.58565784, 0.47053094]),
 array([0.60975989, 0.27418549, 0.82129089, 0.65671476, 0.73571342,
        0.89617115, 1.17445006, 0.60660711, 0.41801673, 0.97014978]))

#### 45. 创建一个长度为10的随机向量，并将其中最大的数改为0

In [95]:
z = np.random.random(10)
z

array([0.40222697, 0.24585117, 0.98026263, 0.69443148, 0.00198175,
       0.70167767, 0.64086571, 0.32269992, 0.36033104, 0.23438615])

In [96]:
z[z.argmax()] = 0
z

array([0.40222697, 0.24585117, 0.        , 0.69443148, 0.00198175,
       0.70167767, 0.64086571, 0.32269992, 0.36033104, 0.23438615])

#### 46. 创建一个结构化的数组，其元素为x轴，y轴的座标，并覆盖[0,1]x[0, 1]

In [97]:
z = np.zeros((5, 5), [('x', float), ('y', float)])
z = np.meshgrid(np.linspace(0, 1, 5),
                np.linspace(0, 1, 5))
z

(array([[0.  , 0.25, 0.5 , 0.75, 1.  ],
        [0.  , 0.25, 0.5 , 0.75, 1.  ],
        [0.  , 0.25, 0.5 , 0.75, 1.  ],
        [0.  , 0.25, 0.5 , 0.75, 1.  ],
        [0.  , 0.25, 0.5 , 0.75, 1.  ]]),
 array([[0.  , 0.  , 0.  , 0.  , 0.  ],
        [0.25, 0.25, 0.25, 0.25, 0.25],
        [0.5 , 0.5 , 0.5 , 0.5 , 0.5 ],
        [0.75, 0.75, 0.75, 0.75, 0.75],
        [1.  , 1.  , 1.  , 1.  , 1.  ]]))

#### 47. 给定两个数组X，Y， 计算其柯西矩阵C(Cauchy Matrix)并求其行列式
$$C_{ij} = \frac{1}{x_i-y_j}$$

In [98]:
x = np.arange(8)
y = x + 0.5
C = 1 / np.subtract.outer(x, y)
C

array([[-2.        , -0.66666667, -0.4       , -0.28571429, -0.22222222,
        -0.18181818, -0.15384615, -0.13333333],
       [ 2.        , -2.        , -0.66666667, -0.4       , -0.28571429,
        -0.22222222, -0.18181818, -0.15384615],
       [ 0.66666667,  2.        , -2.        , -0.66666667, -0.4       ,
        -0.28571429, -0.22222222, -0.18181818],
       [ 0.4       ,  0.66666667,  2.        , -2.        , -0.66666667,
        -0.4       , -0.28571429, -0.22222222],
       [ 0.28571429,  0.4       ,  0.66666667,  2.        , -2.        ,
        -0.66666667, -0.4       , -0.28571429],
       [ 0.22222222,  0.28571429,  0.4       ,  0.66666667,  2.        ,
        -2.        , -0.66666667, -0.4       ],
       [ 0.18181818,  0.22222222,  0.28571429,  0.4       ,  0.66666667,
         2.        , -2.        , -0.66666667],
       [ 0.15384615,  0.18181818,  0.22222222,  0.28571429,  0.4       ,
         0.66666667,  2.        , -2.        ]])

In [99]:
np.linalg.det(C)

np.float64(3638.163637117973)

其实这里`np.subtract.outer`就等于进行了broadcast,我们也可以像下面这样写。

In [100]:
C_test = 1 / (x.reshape(8, 1)- y.reshape(1, 8))
C_test

array([[-2.        , -0.66666667, -0.4       , -0.28571429, -0.22222222,
        -0.18181818, -0.15384615, -0.13333333],
       [ 2.        , -2.        , -0.66666667, -0.4       , -0.28571429,
        -0.22222222, -0.18181818, -0.15384615],
       [ 0.66666667,  2.        , -2.        , -0.66666667, -0.4       ,
        -0.28571429, -0.22222222, -0.18181818],
       [ 0.4       ,  0.66666667,  2.        , -2.        , -0.66666667,
        -0.4       , -0.28571429, -0.22222222],
       [ 0.28571429,  0.4       ,  0.66666667,  2.        , -2.        ,
        -0.66666667, -0.4       , -0.28571429],
       [ 0.22222222,  0.28571429,  0.4       ,  0.66666667,  2.        ,
        -2.        , -0.66666667, -0.4       ],
       [ 0.18181818,  0.22222222,  0.28571429,  0.4       ,  0.66666667,
         2.        , -2.        , -0.66666667],
       [ 0.15384615,  0.18181818,  0.22222222,  0.28571429,  0.4       ,
         0.66666667,  2.        , -2.        ]])

In [101]:
np.linalg.det(C_test)

np.float64(3638.163637117973)

In [102]:
# 测试两种方法返回的C是否相同
np.array_equal(C, C_test)

True

#### 48. 打印Numpy所有标量类型(scalar type)可表示的最值

In [103]:
for dtype in [np.int8, np.int16, np.int32, np.int64]:
    info = np.iinfo(dtype)
    print(f"{dtype}: min={info.min}, max={info.max}")
for dtype in [np.float16, np.float32, np.float64,]:
    info = np.finfo(dtype)
    print(f"{dtype}: min={info.min}, max={info.max}")

<class 'numpy.int8'>: min=-128, max=127
<class 'numpy.int16'>: min=-32768, max=32767
<class 'numpy.int32'>: min=-2147483648, max=2147483647
<class 'numpy.int64'>: min=-9223372036854775808, max=9223372036854775807
<class 'numpy.float16'>: min=-65504.0, max=65504.0
<class 'numpy.float32'>: min=-3.4028234663852886e+38, max=3.4028234663852886e+38
<class 'numpy.float64'>: min=-1.7976931348623157e+308, max=1.7976931348623157e+308


#### 49. 打印数组所有元素（不省略）

In [104]:
with np.printoptions(threshold=np.inf):
    z = np.ones((10, 10))
    print(z)

[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]


#### 50. 给定一个数，在数组中找出距离其最近的数

In [105]:
# 给定的数组
z = np.random.uniform(0, 1, 10)
z

array([0.42325659, 0.95774016, 0.88996072, 0.41034238, 0.37153616,
       0.43800675, 0.15133658, 0.27293758, 0.34080748, 0.80039615])

In [106]:
# 给定的数
x = 0.5
# 定位距离最近数的位置
index = np.abs(z - x).argmin()
# 找到该数字
z[index]

np.float64(0.43800674863239786)

#### 51. 创建一个结构化的数组，其元素为一个座标(x,y)和一个颜色参数(r,g,b)
和46题类似。
>另外，这可以是一个像素点的表示方式

In [107]:
z = np.zeros(10, [('position', [('x', float, 1),
                               ('y', float, 1)]),
                 ('color', [('r', float, 1),
                            ('g', float, 1),
                            ('b', float, 1)])
                 ]
            )
z

array([(([0.], [0.]), ([0.], [0.], [0.])),
       (([0.], [0.]), ([0.], [0.], [0.])),
       (([0.], [0.]), ([0.], [0.], [0.])),
       (([0.], [0.]), ([0.], [0.], [0.])),
       (([0.], [0.]), ([0.], [0.], [0.])),
       (([0.], [0.]), ([0.], [0.], [0.])),
       (([0.], [0.]), ([0.], [0.], [0.])),
       (([0.], [0.]), ([0.], [0.], [0.])),
       (([0.], [0.]), ([0.], [0.], [0.])),
       (([0.], [0.]), ([0.], [0.], [0.]))],
      dtype=[('position', [('x', '<f8', (1,)), ('y', '<f8', (1,))]), ('color', [('r', '<f8', (1,)), ('g', '<f8', (1,)), ('b', '<f8', (1,))])])

#### 52. 考虑一个形状为（10， 2）的随机向量，若其代表二维平面中的点，求各点之间的距离

In [108]:
z = np.random.random((10, 2))
x, y = np.atleast_2d(z[:, 0], z[:, 1])
d = np.sqrt((x - x.T)**2 + (y - y.T)**2)
d

array([[0.        , 0.28021539, 0.69225706, 0.12226023, 0.2210915 ,
        0.66879251, 0.37277058, 0.6740914 , 0.58308322, 0.89960334],
       [0.28021539, 0.        , 0.81910351, 0.28297719, 0.13949434,
        0.94892058, 0.3904595 , 0.80877732, 0.86312854, 0.99143516],
       [0.69225706, 0.81910351, 0.        , 0.81400481, 0.86582721,
        0.79381183, 0.44261454, 0.02766022, 0.74240291, 0.22363469],
       [0.12226023, 0.28297719, 0.81400481, 0.        , 0.16909318,
        0.70636292, 0.48302486, 0.79547951, 0.62345922, 1.02186102],
       [0.2210915 , 0.13949434, 0.86582721, 0.16909318, 0.        ,
        0.87063065, 0.46910365, 0.8515875 , 0.78653903, 1.05610601],
       [0.66879251, 0.94892058, 0.79381183, 0.70636292, 0.87063065,
        0.        , 0.85279972, 0.76625542, 0.0860729 , 0.99723163],
       [0.37277058, 0.3904595 , 0.44261454, 0.48302486, 0.46910365,
        0.85279972, 0.        , 0.43724332, 0.77240531, 0.60141801],
       [0.6740914 , 0.80877732, 0.0276602

这里使用`np.atleast_2d`使得我们得到的`x`， `y`直接就是2维的数组，方便了我们后面直接使用`broadcasting`.我们也可以采用下面的方法代替这行，但是不够简洁：
```python
x = z[:, 0].reshape(10, 1)
y = z[:, 1].reshape(1, 10)
```

此外我们也可以使用`scipy`内置的函数，其效率要高一些。

In [109]:
import scipy
import scipy.spatial

d = scipy.spatial.distance.cdist(z, z)
d

array([[0.        , 0.28021539, 0.69225706, 0.12226023, 0.2210915 ,
        0.66879251, 0.37277058, 0.6740914 , 0.58308322, 0.89960334],
       [0.28021539, 0.        , 0.81910351, 0.28297719, 0.13949434,
        0.94892058, 0.3904595 , 0.80877732, 0.86312854, 0.99143516],
       [0.69225706, 0.81910351, 0.        , 0.81400481, 0.86582721,
        0.79381183, 0.44261454, 0.02766022, 0.74240291, 0.22363469],
       [0.12226023, 0.28297719, 0.81400481, 0.        , 0.16909318,
        0.70636292, 0.48302486, 0.79547951, 0.62345922, 1.02186102],
       [0.2210915 , 0.13949434, 0.86582721, 0.16909318, 0.        ,
        0.87063065, 0.46910365, 0.8515875 , 0.78653903, 1.05610601],
       [0.66879251, 0.94892058, 0.79381183, 0.70636292, 0.87063065,
        0.        , 0.85279972, 0.76625542, 0.0860729 , 0.99723163],
       [0.37277058, 0.3904595 , 0.44261454, 0.48302486, 0.46910365,
        0.85279972, 0.        , 0.43724332, 0.77240531, 0.60141801],
       [0.6740914 , 0.80877732, 0.0276602

#### 53. 将一个32位的浮点数数组，（不使用额外内存）转化为32为的整数数组

In [110]:
z = np.zeros(10, dtype=np.float32)
z

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.], dtype=float32)

In [111]:
z = z.astype(np.int32, copy=False)
z

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

#### 54. 如何读取下面的文件
```
1, 2, 3, 4, 5
6,  ,  , 7, 8
 ,  , 9,10,11
```

In [112]:
from io import StringIO
# “假”的文件
s = StringIO("""1, 2, 3, 4, 5\n
                6,  ,  , 7, 8\n
                 ,  , 9,10,11\n""")
z = np.genfromtxt(s, delimiter=",", missing_values=' ')
z

array([[ 1.,  2.,  3.,  4.,  5.],
       [ 6., nan, nan,  7.,  8.],
       [nan, nan,  9., 10., 11.]])

#### 55. Python内置有`enumerate`,Numpy中与之对应的是？

In [113]:
Z = np.arange(9).reshape(3,3)
for index, value in np.ndenumerate(Z):
    print(index, value)

(0, 0) 0
(0, 1) 1
(0, 2) 2
(1, 0) 3
(1, 1) 4
(1, 2) 5
(2, 0) 6
(2, 1) 7
(2, 2) 8


In [114]:
for index in np.ndindex(Z.shape):
    print(index, Z[index])

(0, 0) 0
(0, 1) 1
(0, 2) 2
(1, 0) 3
(1, 1) 4
(1, 2) 5
(2, 0) 6
(2, 1) 7
(2, 2) 8


#### 56. 生成二维高斯分布

In [115]:
X, Y = np.meshgrid(np.linspace(-1,1,10), np.linspace(-1,1,10))
D = np.sqrt(X*X+Y*Y)
sigma, mu = 1.0, 0.0
G = np.exp(-( (D-mu)**2 / ( 2.0 * sigma**2 ) ) )
G

array([[0.36787944, 0.44822088, 0.51979489, 0.57375342, 0.60279818,
        0.60279818, 0.57375342, 0.51979489, 0.44822088, 0.36787944],
       [0.44822088, 0.54610814, 0.63331324, 0.69905581, 0.73444367,
        0.73444367, 0.69905581, 0.63331324, 0.54610814, 0.44822088],
       [0.51979489, 0.63331324, 0.73444367, 0.81068432, 0.85172308,
        0.85172308, 0.81068432, 0.73444367, 0.63331324, 0.51979489],
       [0.57375342, 0.69905581, 0.81068432, 0.89483932, 0.9401382 ,
        0.9401382 , 0.89483932, 0.81068432, 0.69905581, 0.57375342],
       [0.60279818, 0.73444367, 0.85172308, 0.9401382 , 0.98773022,
        0.98773022, 0.9401382 , 0.85172308, 0.73444367, 0.60279818],
       [0.60279818, 0.73444367, 0.85172308, 0.9401382 , 0.98773022,
        0.98773022, 0.9401382 , 0.85172308, 0.73444367, 0.60279818],
       [0.57375342, 0.69905581, 0.81068432, 0.89483932, 0.9401382 ,
        0.9401382 , 0.89483932, 0.81068432, 0.69905581, 0.57375342],
       [0.51979489, 0.63331324, 0.7344436

#### 57. 随机地在二维数组中放置p个元素

In [116]:
n = 5
p = 3
z = np.zeros((n, n))
np.put(z, np.random.choice(range(n*n), p), 1)
z

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 0., 1.],
       [0., 0., 1., 0., 0.]])

#### 58. 矩阵每行进行中心化（减去均值）

In [117]:
# 1, 答案的方法
z = np.arange(10).reshape(2, 5)
z

array([[0, 1, 2, 3, 4],
       [5, 6, 7, 8, 9]])

In [118]:
z_new = z - z.mean(axis=1, keepdims=True)
z_new

array([[-2., -1.,  0.,  1.,  2.],
       [-2., -1.,  0.,  1.,  2.]])

注意这里，设置`keepdims`可以方便进行`broadcasting`, 免去手动`reshape`的流程。

我们也可以考虑对矩阵每一行应用一个中心化的函数来完成任务。

In [119]:
# 2, apply方法
def centered(xs):
    return xs - xs.mean()

z_new = np.apply_along_axis(centered, 1, z)
z_new

array([[-2., -1.,  0.,  1.,  2.],
       [-2., -1.,  0.,  1.,  2.]])

#### 59. 根据某列数据来排列数组

In [120]:
z = np.random.randint(0, 10, (3, 3))
z

array([[0, 1, 9],
       [7, 6, 7],
       [1, 8, 4]])

In [121]:
# 根据第二列顺序排列
z[z[:, 1].argsort(), ]

array([[0, 1, 9],
       [7, 6, 7],
       [1, 8, 4]])

#### 60. 判断二维数组是否含有空列(全为0)

In [122]:
z = np.random.randint(0, 3, (3, 10))
z

array([[2, 0, 2, 1, 2, 1, 2, 1, 0, 0],
       [2, 0, 1, 1, 2, 2, 1, 1, 1, 1],
       [2, 0, 0, 0, 2, 1, 2, 2, 2, 0]])

In [123]:
print((~z.any(axis=0)).any())

True


一旦有空列的时候， `z.any(axis=0)`返回`False`， 即`~z.any(axis=0)`返回`True`， 之后再应用`any`, 则比返回`True`。 反之，若无任何空列，`~z.any(axis=0)`全部返回`False`， 应用`any`，依旧返回`False`。

#### 61. 给定一个数，在数组中找出距离其最近的数
与50题重复.

#### 62. 考虑两个数组，形状分别是（3， 1）， （1， 3），如何使用迭代器将其相加？

In [124]:
A = np.arange(3).reshape(3,1)
B = np.arange(3).reshape(1,3)
it = np.nditer([A,B,None])
for x,y,z in it:
    z[...] = x + y
print(it.operands[2])

[[0 1 2]
 [1 2 3]
 [2 3 4]]


#### 63. 创建一个带有名称属性的数组类

In [125]:
class NamedArray(np.ndarray):
    def __new__(cls, array, name="no name"):
        obj = np.asarray(array).view(cls)
        obj.name = name
        return obj
    def __array_finalize__(self, obj):
        if obj is None:
            return
        self.info = getattr(obj, 'name', "no name")

Z = NamedArray(np.arange(10), "range_10")
print (Z.name)

range_10


#### 64. 给定一个数值向量，和一个索引向量，根据后者的索引，在前者对应位置加1（注意重复索引）

In [126]:
# Author: Brett Olsen
Z = np.ones(10)
indices = np.random.randint(0,len(Z),20)

Z_new = Z + np.bincount(indices, minlength=len(Z))
Z_new

array([2., 1., 3., 4., 5., 4., 3., 5., 1., 2.])

In [127]:
# Another solution
# Author: Bartosz Telenczuk
np.add.at(Z, indices, 1)
Z

array([2., 1., 3., 4., 5., 4., 3., 5., 1., 2.])

#### 65. 根据索引列表I(indices)，对数值列表X进行累加，得到F

In [128]:
X = [1,2,3,4,5,6]
indices = [1,3,9,3,4,1]
F = np.bincount(indices,X)
print(F)

[0. 7. 0. 6. 5. 0. 0. 0. 0. 3.]


这里`bincount`的用法有点绕...让我们举个例子先:-)

我们把`I`中出现的数字比作个人的银行账户编号，可以看到这里最大的编号为9，所以我们暂时可以只考虑编号0~9的账户情况， 这10个账户，正好对应最后得到F的10个位置。进一步地，`I`与`X`结合，可以看作这些银行账户交易的流水，其中`I`为账户编号，`X`为对应的金额。比如，因为`I[0]=1`，我们知道是账户1发生交易，对应的`X[0]=1`，所以账户1的金额要加1；此外账户1还发生一次交易（`I[5]=1`）, 对应的金额`X[5] = 6`，所以这段时间账户1总的金额就是`6 + 1 = 7`，所以得到`F[1] = 7`.

简言之，我们的任务就是根据`X`和`I`组成的交易流水，来计算各个账户总的金额。

#### 66. 给定一张照片（w, h, 3），计算其中不同颜色的个数

In [129]:
# Author: Fisher Wang

w, h = 256, 256
indices = np.random.randint(0, 4, (h, w, 3)).astype(np.ubyte)
colors = np.unique(indices.reshape(-1, 3), axis=0)
n = len(colors)
print(n)

64


#### 67. 考虑一个四维数组，计算后两个轴上的元素和

In [130]:
A = np.random.randint(0,10,(3,4,3,4))
A.sum(axis=(-2,-1))

array([[57, 58, 46, 75],
       [41, 58, 51, 44],
       [53, 64, 55, 36]])

#### 68. 给定向量D，根据索引数组S得到子集，计算子集上的均值

In [131]:
D = np.random.uniform(0,1,100)
S = np.random.randint(0,10,100)
D_sums = np.bincount(S, weights=D)
D_counts = np.bincount(S)
D_means = D_sums / D_counts
print(D_means)

[0.47583679 0.39951166 0.47921084 0.49696441 0.53094028 0.47477069
 0.46135421 0.60496636 0.46196741 0.53962136]


结合65题给出的例子，在那里是根据流水计算各个账户总的金额，这里是计算各个账户每次交易的平均金额，也就是该账户总的金额除以其交易的次数。

#### 69. 获取矩阵点乘(dot product)结果的对角线元素

In [132]:
# Author: Mathieu Blondel

A = np.random.uniform(0,1,(5,5))
B = np.random.uniform(0,1,(5,5))

# 慢的版本
np.diag(np.dot(A, B))

# 快的版本
np.sum(A * B.T, axis=1)

# 更快的版本
np.einsum("ij,ji->i", A, B)

array([1.03321594, 1.3897372 , 2.18677728, 1.9050502 , 1.34570535])

#### 70. 如何在数组[1, 2, 3, 4, 5]每两个值的中间添加三个0

In [133]:
z = np.array([1, 2, 3, 4, 5])
nz = 3  # 0的个数
v = np.zeros(len(z) + nz*(len(z)-1))
v[::nz+1] = z
v

array([1., 0., 0., 0., 2., 0., 0., 0., 3., 0., 0., 0., 4., 0., 0., 0., 5.])

#### 71. 维度分别为（5， 5， 3）， （5， 5）的两个数组相乘

In [134]:
A = np.ones((5,5,3))
B = 2*np.ones((5,5))
print(A * B[:,:,None])

[[[2. 2. 2.]
  [2. 2. 2.]
  [2. 2. 2.]
  [2. 2. 2.]
  [2. 2. 2.]]

 [[2. 2. 2.]
  [2. 2. 2.]
  [2. 2. 2.]
  [2. 2. 2.]
  [2. 2. 2.]]

 [[2. 2. 2.]
  [2. 2. 2.]
  [2. 2. 2.]
  [2. 2. 2.]
  [2. 2. 2.]]

 [[2. 2. 2.]
  [2. 2. 2.]
  [2. 2. 2.]
  [2. 2. 2.]
  [2. 2. 2.]]

 [[2. 2. 2.]
  [2. 2. 2.]
  [2. 2. 2.]
  [2. 2. 2.]
  [2. 2. 2.]]]


#### 72. 交换数组的两行

In [135]:
A = np.arange(25).reshape((5, 5))
A

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24]])

In [136]:
# 交换第一、二两行
A[[0, 1]] = A[[1, 0]]
A

array([[ 5,  6,  7,  8,  9],
       [ 0,  1,  2,  3,  4],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24]])

#### 73. 给定10个三元组描述10个三角形，找出所有边的集合

In [137]:
faces = np.random.randint(0,100,(10,3))
F = np.roll(faces.repeat(2,axis=1),-1,axis=1)
F = F.reshape(len(F)*3,2)
F = np.sort(F,axis=1)
G = F.view( dtype=[('p0',F.dtype),('p1',F.dtype)] )
G = np.unique(G)
print(G)

[( 0, 44) ( 0, 59) ( 5, 61) ( 5, 81) (14, 21) (14, 29) (15, 28) (15, 85)
 (16, 53) (16, 83) (21, 25) (21, 29) (21, 86) (23, 59) (23, 61) (25, 86)
 (28, 31) (28, 85) (28, 97) (31, 97) (34, 62) (34, 75) (43, 45) (43, 80)
 (44, 59) (45, 80) (53, 83) (59, 61) (61, 81) (62, 75)]


感觉这里的方法比较巧妙，可以将每一步拆解来理解怎么将每条边抽取出来。之后比较细节的地方就是对描述“边”的二元组排序(如果不排序，后面比较的时候就会出现(a, b)与（b, a）不是同一条边的错误判断)，之后通过`view`转化类型，方便比较（使用`np.unique`）

#### 74. 给定A，我们有C = np.bincount(A), 那么，给定C，如何找到对应的A？

In [138]:
C = np.bincount([1,1,2,3,4,4,6])
A = np.repeat(np.arange(len(C)), C)
print(A)

[1 1 2 3 4 4 6]


#### 75. 用滑动窗口计算平均值

In [139]:
def moving_average(a, n=3) :
    ret = np.cumsum(a, dtype=float)
    ret[n:] = ret[n:] - ret[:-n]
    return ret[n - 1:] / n
Z = np.arange(20)
print(moving_average(Z, n=3))

[ 1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16. 17. 18.]


#### 76. 给定一个一维数组，组建一个二维数组，使得第一行为Z[0], Z[1], Z[2], 第二行为Z[1], Z[2], Z[3], 依此类推，最后一行为Z[-3], Z[-2], Z[-1]

In [140]:
from numpy.lib import stride_tricks

def rolling(a, window):
    shape = (a.size - window + 1, window)
    strides = (a.itemsize, a.itemsize)
    return stride_tricks.as_strided(a, shape=shape, strides=strides)
Z = rolling(np.arange(10), 3)
print(Z)

[[0 1 2]
 [1 2 3]
 [2 3 4]
 [3 4 5]
 [4 5 6]
 [5 6 7]
 [6 7 8]
 [7 8 9]]


#### 77. 如何原地对布尔值取反，如何原地改变数字的正负

In [141]:
Z = np.random.randint(0,2,100)
np.logical_not(Z, out=Z)

Z = np.random.uniform(-1.0,1.0,100)
np.negative(Z, out=Z)

array([ 0.15153121,  0.10326763,  0.38301685, -0.46161341,  0.05677491,
       -0.92037266,  0.86484754, -0.43024656,  0.08460352,  0.49146569,
       -0.24792258,  0.2684628 ,  0.86281593, -0.52276367,  0.83281867,
        0.61442957,  0.01925966,  0.0291909 , -0.72259418,  0.2551861 ,
       -0.63864825, -0.29881944,  0.03796013,  0.74960219, -0.64582809,
        0.3711555 , -0.38367882, -0.26371576,  0.91023875,  0.92276948,
       -0.93524825, -0.83931958, -0.20283998,  0.56130221,  0.38103203,
       -0.32978579, -0.07731042,  0.8497525 , -0.08654566, -0.82750775,
       -0.78896552, -0.77512467, -0.45471949, -0.59343759,  0.83760065,
        0.28339496, -0.64119305, -0.18546573, -0.23683516,  0.0454317 ,
       -0.74016927,  0.62682431, -0.58787537,  0.18017752,  0.63161805,
        0.82974425,  0.4673166 , -0.63450752, -0.45985772,  0.10885944,
       -0.40991301,  0.94869332, -0.86108914,  0.71112426,  0.94251363,
       -0.31550912, -0.57789077, -0.9943572 ,  0.74751539,  0.65

#### 78. 计算点p到各个直线i的距离, 其中直线由(P0[i]， P1[i])表示，P0，P1为一系列对应的点

In [142]:
def distance(P0, P1, p):
    T = P1 - P0
    L = (T**2).sum(axis=1)
    U = -((P0[:,0]-p[...,0])*T[:,0] + (P0[:,1]-p[...,1])*T[:,1]) / L
    U = U.reshape(len(U),1)
    D = P0 + U*T - p
    return np.sqrt((D**2).sum(axis=1))

P0 = np.random.uniform(-10,10,(10,2))
P1 = np.random.uniform(-10,10,(10,2))
p  = np.random.uniform(-10,10,( 1,2))
print(distance(P0, P1, p))

[ 0.67014654  8.79415464  5.19016452 12.83345647  5.15295786  4.10113663
  3.81029025  5.8810243   1.09608287  7.03041337]


#### 79. 接上题，如何计算P0中各点P0[j]到各直线（P0[i], P1[i]）的距离

In [143]:
# based on distance function from previous question
P0 = np.random.uniform(-10, 10, (10,2))
P1 = np.random.uniform(-10,10,(10,2))
p = np.random.uniform(-10, 10, (10,2))
print(np.array([distance(P0,P1,p_i) for p_i in p]))

[[ 0.53747476  1.79873769  3.05363745  2.11170594 10.17753139  8.1058492
   4.1530989   8.95744069  7.03315705 13.66240662]
 [ 1.17901004  1.75345754  4.65415474  1.42124792 12.08923837  9.65047045
   5.15760066 10.88812951  7.6945859  15.41636134]
 [ 6.41191744  4.22922729  2.72887767  6.20193212  4.48642472  5.15155288
   3.30044673  3.56681884  7.34067451  9.70126649]
 [ 2.91166044  5.60966409  1.84108961  2.60412236  4.73780193  1.6869625
   5.72380404  2.29161355  2.3234107   4.53358746]
 [ 3.59130825 14.61830461  9.0890202  11.92769948  8.67370352  4.89139434
  11.95764335  4.88019713  9.80281114  3.25011145]
 [ 9.75631389  4.31035322  5.84637198  7.54183194  0.75900179  2.13474123
   1.33521344  0.19872819  6.04392751  6.27749571]
 [ 3.40583685 14.14478077  2.52588676  8.89376967  0.98886384 10.89183922
  15.72417888  2.83561696 12.17260256  3.64054652]
 [ 2.62003644  7.73283054  2.39651781  4.32716749  4.27102321  3.59248284
   7.95643911  1.48831616  4.60136167  2.9336322 ]
 [

#### 80. 给定任意一个数组，编写一个函数，接受数组和一个元素为参数，返回以元素为中心的子集（必要的时候可以进行填充）

In [144]:
# Author: Nicolas Rougier

Z = np.random.randint(0,10,(10,10))
shape = (5,5)
fill  = 0
position = (1,1)

R = np.ones(shape, dtype=Z.dtype)*fill
P  = np.array(list(position)).astype(int)
Rs = np.array(list(shape)).astype(int)
Zs = np.array(list(Z.shape)).astype(int)

R_start = np.zeros((len(shape),)).astype(int)
R_stop  = np.array(list(shape)).astype(int)
Z_start = (P - Rs//2)
Z_stop  = (P + Rs//2) + Rs%2

R_start = (R_start - np.minimum(Z_start, 0)).tolist()
Z_start = np.maximum(Z_start, 0).tolist()
R_stop = np.maximum(R_start, (R_stop - np.maximum(Z_stop - Zs, 0))).tolist()
Z_stop = np.minimum(Z_stop, Zs).tolist()

# Convert slices to tuples for proper indexing
r = tuple([slice(start, stop) for start, stop in zip(R_start, R_stop)])
z = tuple([slice(start, stop) for start, stop in zip(Z_start, Z_stop)])

R[r] = Z[z]
print(Z)
print(R)

[[7 7 7 4 8 2 2 7 4 8]
 [2 8 7 1 9 3 2 3 8 4]
 [5 1 9 3 6 7 5 0 9 0]
 [4 5 0 4 3 8 1 3 4 7]
 [9 0 3 0 4 0 0 4 3 0]
 [3 0 6 5 1 6 3 4 8 1]
 [5 4 2 2 6 7 0 4 4 7]
 [9 1 7 9 9 4 9 1 6 6]
 [0 3 1 0 6 5 7 9 7 0]
 [9 0 2 8 1 8 4 4 6 7]]
[[0 0 0 0 0]
 [0 7 7 7 4]
 [0 2 8 7 1]
 [0 5 1 9 3]
 [0 4 5 0 4]]


PS: 感觉难度为三星的题目（64题及之后）很多出的不太好...考察的是更加灵活地运用Numpy，逻辑是没问题，但是缺乏具体的示例，没有依托实际的问题，就显得比较空洞Orz

#### 81. 有数组Z = [1,2,3,4,5,6,7,8,9,10,11,12,13,14]， 如何生成 = [[1,2,3,4], [2,3,4,5], [3,4,5,6], ..., [11,12,13,14]]？

In [145]:
# 1， 答案的方法
Z = np.arange(1,15,dtype=np.uint32)
R = stride_tricks.as_strided(Z,(11,4),(4,4))
print(R)

[[ 1  2  3  4]
 [ 2  3  4  5]
 [ 3  4  5  6]
 [ 4  5  6  7]
 [ 5  6  7  8]
 [ 6  7  8  9]
 [ 7  8  9 10]
 [ 8  9 10 11]
 [ 9 10 11 12]
 [10 11 12 13]
 [11 12 13 14]]


In [146]:
# 2， 76题的一个特殊形式
z = np.arange(1, 15)
rolling(z, 4)

array([[ 1,  2,  3,  4],
       [ 2,  3,  4,  5],
       [ 3,  4,  5,  6],
       [ 4,  5,  6,  7],
       [ 5,  6,  7,  8],
       [ 6,  7,  8,  9],
       [ 7,  8,  9, 10],
       [ 8,  9, 10, 11],
       [ 9, 10, 11, 12],
       [10, 11, 12, 13],
       [11, 12, 13, 14]])

#### 82. 计算矩阵的秩

In [147]:
# 1， 答案的方法
Z = np.random.uniform(0,1,(10,10))
U, S, V = np.linalg.svd(Z) # Singular Value Decomposition
rank = np.sum(S > 1e-10)
print(rank)

10


In [148]:
# 2， 调用API
np.linalg.matrix_rank(Z)

np.int64(10)

#### 83. 找出数组中的众数

In [149]:
Z = np.random.randint(0,10,50)
Z

array([5, 9, 7, 7, 3, 6, 5, 3, 3, 9, 6, 6, 5, 5, 5, 0, 3, 5, 4, 9, 3, 9,
       3, 4, 3, 8, 9, 3, 9, 4, 0, 8, 2, 9, 2, 3, 0, 8, 7, 5, 4, 5, 1, 7,
       9, 2, 1, 0, 5, 0])

In [150]:
print(np.bincount(Z).argmax())

3


如果不限制在Numpy之内，我们可以直接调用Scipy提供的API

In [151]:
from scipy import stats
stats.mode(Z)

ModeResult(mode=np.int64(3), count=np.int64(9))

#### 84. 从10x10的矩阵中抽取出所有的3x3矩阵

In [152]:
Z = np.random.randint(0,5,(10,10))
n = 3
i = 1 + (Z.shape[0]-3)
j = 1 + (Z.shape[1]-3)
C = stride_tricks.as_strided(Z, shape=(i, j, n, n), strides=Z.strides + Z.strides)
print(C)

[[[[0 1 4]
   [1 4 3]
   [0 3 1]]

  [[1 4 4]
   [4 3 3]
   [3 1 2]]

  [[4 4 3]
   [3 3 4]
   [1 2 0]]

  [[4 3 4]
   [3 4 4]
   [2 0 3]]

  [[3 4 0]
   [4 4 3]
   [0 3 4]]

  [[4 0 1]
   [4 3 0]
   [3 4 1]]

  [[0 1 2]
   [3 0 0]
   [4 1 0]]

  [[1 2 0]
   [0 0 2]
   [1 0 3]]]


 [[[1 4 3]
   [0 3 1]
   [0 4 3]]

  [[4 3 3]
   [3 1 2]
   [4 3 3]]

  [[3 3 4]
   [1 2 0]
   [3 3 3]]

  [[3 4 4]
   [2 0 3]
   [3 3 2]]

  [[4 4 3]
   [0 3 4]
   [3 2 4]]

  [[4 3 0]
   [3 4 1]
   [2 4 2]]

  [[3 0 0]
   [4 1 0]
   [4 2 3]]

  [[0 0 2]
   [1 0 3]
   [2 3 2]]]


 [[[0 3 1]
   [0 4 3]
   [1 4 4]]

  [[3 1 2]
   [4 3 3]
   [4 4 2]]

  [[1 2 0]
   [3 3 3]
   [4 2 2]]

  [[2 0 3]
   [3 3 2]
   [2 2 0]]

  [[0 3 4]
   [3 2 4]
   [2 0 4]]

  [[3 4 1]
   [2 4 2]
   [0 4 3]]

  [[4 1 0]
   [4 2 3]
   [4 3 0]]

  [[1 0 3]
   [2 3 2]
   [3 0 2]]]


 [[[0 4 3]
   [1 4 4]
   [4 2 1]]

  [[4 3 3]
   [4 4 2]
   [2 1 1]]

  [[3 3 3]
   [4 2 2]
   [1 1 1]]

  [[3 3 2]
   [2 2 0]
   [1 1 4]]

  [[3 2 4]
   

#### 85. 构造二维数组的子类，使得Z[i, j] = Z[j, i]

In [153]:
# Author: Eric O. Lebigot
# Note: only works for 2d array and value setting using indices

class Symetric(np.ndarray):
    def __setitem__(self, index, value):
        i,j = index
        super(Symetric, self).__setitem__((i,j), value)
        super(Symetric, self).__setitem__((j,i), value)

def symetric(Z):
    return np.asarray(Z + Z.T - np.diag(Z.diagonal())).view(Symetric)

S = symetric(np.random.randint(0,10,(5,5)))
S[2,3] = 42
print(S)

[[ 2 12  6 10 10]
 [12  4  1  7 14]
 [ 6  1  4 42 10]
 [10  7 42  8  6]
 [10 14 10  6  8]]


#### 86. 给定p个(n, n)矩阵和p个(n, 1)向量， 计算张量乘法(tensor product)

In [154]:
p, n = 10, 20
M = np.ones((p,n,n))
V = np.ones((p,n,1))
S = np.tensordot(M, V, axes=[[0, 2], [0, 1]])
print(S)

[[200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]
 [200.]]


#### 87. 给定16x16的数组，将其分成4x4的小块，求每块的和

In [155]:
Z = np.ones((16,16))
k = 4
S = np.add.reduceat(np.add.reduceat(Z, np.arange(0, Z.shape[0], k), axis=0),
                                       np.arange(0, Z.shape[1], k), axis=1)
print(S)

[[16. 16. 16. 16.]
 [16. 16. 16. 16.]
 [16. 16. 16. 16.]
 [16. 16. 16. 16.]]


#### 88. 用数组实现生存游戏

In [156]:
def iterate(Z):
    # Count neighbours
    N = (Z[0:-2,0:-2] + Z[0:-2,1:-1] + Z[0:-2,2:] +
         Z[1:-1,0:-2]                + Z[1:-1,2:] +
         Z[2:  ,0:-2] + Z[2:  ,1:-1] + Z[2:  ,2:])

    # Apply rules
    birth = (N==3) & (Z[1:-1,1:-1]==0)
    survive = ((N==2) | (N==3)) & (Z[1:-1,1:-1]==1)
    Z[...] = 0
    Z[1:-1,1:-1][birth | survive] = 1
    return Z

Z = np.random.randint(0,2,(50,50))
for i in range(100):
    Z = iterate(Z)
print(Z)

[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


#### 89. 获取数组最大的n个值

In [157]:
Z = np.arange(10000)
np.random.shuffle(Z)
n = 5

In [158]:
# 较慢
print (Z[np.argsort(Z)[-n:]])

[9995 9996 9997 9998 9999]


In [159]:
# 较快
print (Z[np.argpartition(-Z,n)[:n]])

[9999 9998 9997 9996 9995]


#### 90. 给定任意大小的数组，计算其笛卡尔积

In [160]:
def cartesian(arrays):
    arrays = [np.asarray(a) for a in arrays]
    shape = (len(x) for x in arrays)

    ix = np.indices(shape, dtype=int)
    ix = ix.reshape(len(arrays), -1).T

    for n, arr in enumerate(arrays):
        ix[:, n] = arrays[n][ix[:, n]]

    return ix

print (cartesian(([1, 2, 3], [4, 5], [6, 7])))

[[1 4 6]
 [1 4 7]
 [1 5 6]
 [1 5 7]
 [2 4 6]
 [2 4 7]
 [2 5 6]
 [2 5 7]
 [3 4 6]
 [3 4 7]
 [3 5 6]
 [3 5 7]]


#### 91. 将一般的数组转化为结构化数组

In [161]:
Z = np.array([("Hello", 2.5, 3),
              ("World", 3.6, 2)])
Z

array([['Hello', '2.5', '3'],
       ['World', '3.6', '2']], dtype='<U32')

In [162]:
R = np.core.records.fromarrays(Z.T, 
                               names='col1, col2, col3',
                               formats = 'S8, f8, i8')
R

/tmp/ipykernel_70371/4252519378.py:1: DeprecationWarning: numpy.core is deprecated and has been renamed to numpy._core. The numpy._core namespace contains private NumPy internals and its use is discouraged, as NumPy internals can change without warning in any release. In practice, most real-world usage of numpy.core is to access functionality in the public NumPy API. If that is the case, use the public NumPy API. If not, you are using NumPy internals. If you would still like to access an internal attribute, use numpy._core.records.
  R = np.core.records.fromarrays(Z.T,


rec.array([(b'Hello', 2.5, 3), (b'World', 3.6, 2)],
          dtype=[('col1', 'S8'), ('col2', '<f8'), ('col3', '<i8')])

#### 92. 用三种方法计算大向量的三次方

In [163]:
x = np.random.rand(int(5e7))

In [164]:
%timeit np.power(x,3)

480 ms ± 1.93 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [165]:
%timeit x*x*x

82.4 ms ± 1.17 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [166]:
%timeit np.einsum('i,i,i->i',x,x,x)

78.5 ms ± 1.34 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


#### 93. 考虑形状（8,3）和（2,2）的两个数组A和B. 如何查找包含B的每一行元素的A行，不考虑B中元素的顺序

In [167]:
A = np.random.randint(0,5,(8,3))
B = np.random.randint(0,5,(2,2))

C = (A[..., np.newaxis, np.newaxis] == B)
rows = np.where(C.any((3,1)).all(1))[0]
print(rows)

[0 2 3 4 5 6]


#### 94. 给定（10， 3）的矩阵， 找出包含不同元素的行

In [168]:
Z = np.random.randint(0,5,(10,3))
Z

array([[1, 2, 2],
       [2, 3, 0],
       [0, 0, 1],
       [3, 2, 4],
       [4, 2, 3],
       [2, 0, 2],
       [0, 4, 4],
       [3, 3, 2],
       [4, 4, 2],
       [4, 3, 1]])

In [169]:
# 适用于任意数据类型
E = np.all(Z[:,1:] == Z[:,:-1], axis=1)
U = Z[~E]
U

array([[1, 2, 2],
       [2, 3, 0],
       [0, 0, 1],
       [3, 2, 4],
       [4, 2, 3],
       [2, 0, 2],
       [0, 4, 4],
       [3, 3, 2],
       [4, 4, 2],
       [4, 3, 1]])

In [170]:
# 仅适用于数值类型
U = Z[Z.max(axis=1) != Z.min(axis=1),:]
U

array([[1, 2, 2],
       [2, 3, 0],
       [0, 0, 1],
       [3, 2, 4],
       [4, 2, 3],
       [2, 0, 2],
       [0, 4, 4],
       [3, 3, 2],
       [4, 4, 2],
       [4, 3, 1]])

#### 95. 将给定的整数向量转化为二进制矩阵

In [171]:
indices = np.array([0, 1, 2, 3, 15, 16, 32, 64, 128], dtype=np.uint8)
print(np.unpackbits(indices[:, np.newaxis], axis=1))

[[0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 1]
 [0 0 0 0 0 0 1 0]
 [0 0 0 0 0 0 1 1]
 [0 0 0 0 1 1 1 1]
 [0 0 0 1 0 0 0 0]
 [0 0 1 0 0 0 0 0]
 [0 1 0 0 0 0 0 0]
 [1 0 0 0 0 0 0 0]]


#### 96. 给定二维数组，抽取所有不同的行

In [172]:
Z = np.random.randint(0,2,(6,3))
Z

array([[1, 1, 1],
       [1, 0, 0],
       [1, 1, 0],
       [1, 1, 0],
       [1, 0, 1],
       [1, 1, 0]])

In [173]:
uZ = np.unique(Z, axis=0)
uZ

array([[1, 0, 0],
       [1, 0, 1],
       [1, 1, 0],
       [1, 1, 1]])

#### 97. 考虑两个数组A，B， 使用np.einsum写出矩阵间的outer, inner, sum, mul函数

In [174]:
A = np.random.uniform(0,1,10)
B = np.random.uniform(0,1,10)

np.einsum('i->', A)       # np.sum(A)
np.einsum('i,i->i', A, B) # A * B
np.einsum('i,i', A, B)    # np.inner(A, B)
np.einsum('i,j->ij', A, B)    # np.outer(A, B)

array([[0.77158631, 0.49330467, 0.80774416, 0.11326308, 0.23287531,
        0.70538756, 0.86148216, 0.14888952, 0.83579325, 0.11818144],
       [0.66072967, 0.42242978, 0.69169259, 0.09699016, 0.19941726,
        0.60404194, 0.73770986, 0.12749802, 0.71571177, 0.10120188],
       [0.62300373, 0.39831014, 0.65219874, 0.09145227, 0.18803105,
        0.56955272, 0.69558855, 0.12021821, 0.67484649, 0.09542352],
       [0.83060264, 0.53103607, 0.86952609, 0.12192623, 0.25068724,
        0.75934054, 0.92737436, 0.16027764, 0.89972058, 0.12722079],
       [0.08499108, 0.05433805, 0.08897391, 0.01247605, 0.02565147,
        0.07769922, 0.09489321, 0.01640035, 0.09206355, 0.01301782],
       [0.56469803, 0.36103307, 0.59116074, 0.08289343, 0.17043359,
        0.51624939, 0.63048977, 0.10896723, 0.61168892, 0.08649302],
       [0.12076824, 0.07721176, 0.12642765, 0.01772787, 0.03644951,
        0.11040685, 0.13483869, 0.0233041 , 0.13081787, 0.01849769],
       [0.54013571, 0.34532944, 0.5654473

#### 98. 用两个向量(X, Y)描述一条轨道，如何对其进行等距抽样

In [175]:
phi = np.arange(0, 10*np.pi, 0.1)
a = 1
x = a*phi*np.cos(phi)
y = a*phi*np.sin(phi)

dr = (np.diff(x)**2 + np.diff(y)**2)**.5 # segment lengths
r = np.zeros_like(x)
r[1:] = np.cumsum(dr)                # integrate path
r_int = np.linspace(0, r.max(), 200) # regular spaced path
x_int = np.interp(r_int, r, x)       # integrate path
y_int = np.interp(r_int, r, y)

#### 99. 给定一个二维数组和一个整数n，提取所有仅包含整数，且元素和为n的行

In [176]:
X = np.asarray([[1.0, 0.0, 3.0, 8.0],
                [2.0, 0.0, 1.0, 1.0],
                [1.5, 2.5, 1.0, 0.0]])
n = 4
M = np.logical_and.reduce(np.mod(X, 1) == 0, axis=-1)
M &= (X.sum(axis=-1) == n)
print(X[M])

[[2. 0. 1. 1.]]


#### 100. 给定一个向量，计算其均值的95%的置信区间

In [177]:
X = np.random.randn(100) 
N = 1000  # 再抽样次数
idx = np.random.randint(0, X.size, (N, X.size))
means = X[idx].mean(axis=1)
confint = np.percentile(means, [2.5, 97.5])
print(confint)

[-0.03873369  0.33764323]
